tests/runtime/test_flow_run.py

```diff
@@ -149,6 +149,38 @@ async def test_tags_pulls_from_api_when_needed(self, monkeypatch, prefect_client
         assert set(flow_run.tags) == {"red", "green"}
 
 
+class TestRunCount:
+    async def test_run_count_is_attribute(self):
+        assert "run_count" in dir(flow_run)
+
+    async def test_run_count_is_zero_when_not_set(self):
+        assert flow_run.run_count == 0
+
+    async def test_run_count_returns_run_count_when_present_dynamically(self):
+        assert flow_run.run_count == 0
+
+        with FlowRunContext.construct(
+            flow_run=FlowRun.construct(id="foo", run_count=10)
+        ):
+            assert flow_run.run_count == 10
+
+        assert flow_run.run_count == 0
+
+    async def test_run_count_from_api(self, monkeypatch, prefect_client):
+        run = await prefect_client.create_flow_run(
+            flow=flow(lambda: None, name="test", retries=5)
+        )
+        assert flow_run.run_count == 0
+
+        await prefect_client.set_flow_run_state(
+            flow_run_id=run.id, state=states.Retrying()
+        )
+
+        monkeypatch.setenv(name="PREFECT__FLOW_RUN_ID", value=str(run.id))
```

Comment:
q: What happens if we don't monkeypatch this? Does the `flow_run.run_count` end up using a different ID?

Reply:
Retrieving the flow run id happens here:

```python
def get_id() -> str:
    flow_run_ctx = FlowRunContext.get()
    task_run_ctx = TaskRunContext.get()
    if flow_run_ctx is not None:
        return str(flow_run_ctx.flow_run.id)
    if task_run_ctx is not None:
        return str(task_run_ctx.task_run.flow_run_id)
    else:
        return os.getenv("PREFECT__FLOW_RUN_ID")
```

Since in this test case we are not in a flow run context or task run context, if we did not set an env var, I believe "PREFECT__FLOW_RUN_ID" would just be None, so we wouldn't be able to check the attribute is correct in that case and `flow_run.run_count` would be 0.

Reply:
That makes sense, thanks!